File: sections/10_greedy_constructive/problems/a_non_decreasing_repairs/solution.py
"""Reference solution for A. Non-Decreasing Repairs."""

import sys


def solve_case(a: list[int]) -> int:
    cost = 0
    need = a[0]
    for x in a[1:]:
        if x < need:
            cost += need - x
        else:
            need = x
    return cost


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        idx += 1
        a = data[idx:idx + n]
        idx += n
        out.append(str(solve_case(a)))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/10_greedy_constructive/problems/a_non_decreasing_repairs/solution.cpp
#include <bits/stdc++.h>
using namespace std;

long long solve_case(const vector<long long> &a) {
    long long cost = 0;
    long long need = a[0];
    for (int i = 1; i < (int)a.size(); ++i) {
        if (a[i] < need) {
            cost += need - a[i];
        } else {
            need = a[i];
        }
    }
    return cost;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n;
        cin >> n;
        vector<long long> a(n);
        for (long long &x : a) cin >> x;
        cout << solve_case(a) << '\n';
    }

    return 0;
}

